wgpu-precompile-macro/src/lib.rs

```diff
@@ -0,0 +1,613 @@
+            source_type,
+            shader_stage,
+            source,
+            entry_point,
+            targets,
+            file_name,
+        })
+    }
+}
+impl PrecompileArgs {
+    fn target_enabled(&self, target: CompileTarget) -> bool {
+        // TODO: only enable if we are actually targeting that platform.
+        // This is especially important for DXIL, which requires dxc. No need
+        // to fail to compile on MacOS if dxc isn't present!
+        match target {
+            CompileTarget::Dxil => self.targets.contains(&CompileTarget::Dxil),
+            CompileTarget::Hlsl => {
+                (self.targets.contains(&CompileTarget::Hlsl)
+                    || self.targets.contains(&CompileTarget::AllSupported))
+                    && !self.targets.contains(&CompileTarget::Dxil)
+            }
+            CompileTarget::AllSupported => unreachable!(),
+            other => {
+                self.targets.contains(&other) || self.targets.contains(&CompileTarget::AllSupported)
+            }
+        }
+    }
+}
+
+/// All fields represented as string literals
+struct PrecompileDxilArgs {
+    hlsl_code: String,
+    entry_point: String,
+    shader_stage: naga::ShaderStage,
+}
+impl Parse for PrecompileDxilArgs {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let hlsl_code = input.parse::<syn::LitStr>()?.value();
+        let entry_point = input.parse::<syn::LitStr>()?.value();
+        let shader_stage = parse_shader_stage(&input.parse::<syn::LitStr>()?.value()).unwrap();
+        Ok(Self {
+            hlsl_code,
+            entry_point,
+            shader_stage,
+        })
+    }
+}
+
+/// This is so we can conditionally hook into DXC depending on the target, which must be configured via #\[cfg] within the main program.
+/// Proc macros don't directly have access to the target configuration. Invoking naga unnecessarily shouldn't be *too* major, but
+/// compiling to macOS shouldn't require dxc be present.
+#[proc_macro]
+pub fn precompile_hlsl_to_dxil(input: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(input as PrecompileDxilArgs);
+    let target_profile = match args.shader_stage {
+        naga::ShaderStage::Vertex => "vs_5_1",
+        naga::ShaderStage::Fragment => "ps_5_1",
+        naga::ShaderStage::Compute => "cs_5_1",
+        naga::ShaderStage::Task => "as_5_1",
+        naga::ShaderStage::Mesh => "ms_5_1",
+    };
+
+    let temporary_folder_location = std::env::temp_dir().join(
+        getrandom::u64()
+            .expect("Failed to generate random u64")
+            .to_string(),
+    );
+    std::fs::create_dir(&temporary_folder_location).expect("Failed to create temporary directory");
+
+    // The naming matters for DXIL debug info. We don't want to give it a name that seems like something the
+    // user might've specified, as that could cause confusion.
+    let input_file = temporary_folder_location.join("__wgpu_inline.hlsl");
```

Comment:
Definitely need some kind of guard so that if DXC fails the temp directory is removed.